src/object/srv_ec.c

```diff
@@ -0,0 +1,296 @@
+ * limitations under the License.
+ "
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * This file is part of daos_sr
+ *
+ * src/object/srv_ec.c
+ */
+#define D_LOGFAC	DD_FAC(object)
+
+#include <stddef.h>
+#include <stdio.h>
+#include <daos/rpc.h>
+#include <daos_types.h>
+#include "obj_rpc.h"
+#include "obj_internal.h"
+
+/* Determines if entire update affects just this target. If so, no IOD
+ * modifications or special bulk-transfer handling are needed for this
+ * update request.
+ */
+static bool
+ec_is_one_cell(daos_iod_t *iod, struct daos_oclass_attr *oca,
+	       unsigned int tgt_idx)
+{
+	unsigned int    len = oca->u.ec.e_len;
+	unsigned int    k = oca->u.ec.e_k;
+	bool		rc;
+	unsigned int	j;
+
+	for (j = 0; j < iod->iod_nr; j++) {
+		daos_recx_t     *this_recx = &iod->iod_recxs[j];
+		uint64_t         recx_start_offset = this_recx->rx_idx *
+						     iod->iod_size;
+		uint64_t         recx_end_offset =
+					(this_recx->rx_nr * iod->iod_size) +
+					recx_start_offset;
+
+		if (recx_start_offset & PARITY_INDICATOR) {
+			rc = false;
+			break;
+		} else if (recx_start_offset/len == recx_end_offset/len &&
+			(recx_start_offset % (len * k)) / len == tgt_idx) {
+			rc = true;
+		} else {
+			rc = false;
+			break;
+		}
+	}
+	return rc;
+}
+
+/* Removes the specified recx entry from an IOD's iod_recx array.
+*/
+static void
+ec_del_recx(daos_iod_t *iod, unsigned int idx)
+{
+	int j;
+
+	for (j = idx; j < iod->iod_nr - 1; j++)
+		iod->iod_recxs[j] = iod->iod_recxs[j+1];
+	iod->iod_nr--;
+}
+
+/* Process IOD array on data target. Keeps extents that are addressed
+ * to this target.
+ */
+int
+ec_data_target(unsigned int dtgt_idx, unsigned int nr, daos_iod_t *iods,
+	       struct daos_oclass_attr *oca, long **skip_list)
+{
+	unsigned long	ss = oca->u.ec.e_len * oca->u.ec.e_k;
+	unsigned int	i, j, idx;
+	int		rc = 0;
+
+	for (i = 0; i < nr; i++) {
+		daos_iod_t	*iod = &iods[i];
+		unsigned int	 loop_bound = iod->iod_nr;
+		int		 sl_idx = 0;
+
+		if (iod->iod_type == DAOS_IOD_SINGLE ||
+			ec_is_one_cell(iod, oca, dtgt_idx))
+			continue;
+		D_ALLOC_ARRAY(skip_list[i], 3 * iod->iod_nr + 1);
+		if (skip_list[i] == NULL)
+			D_GOTO(out, rc = -1);
+		for (idx = 0, j = 0; j < loop_bound; j++) {
+			daos_recx_t	*this_recx = &iod->iod_recxs[idx];
+			unsigned long	so =
+				(this_recx->rx_idx * iod->iod_size) % ss;
+			unsigned int	cell = so / oca->u.ec.e_len;
+			unsigned long	recx_size = iod->iod_size *
+						 this_recx->rx_nr;
+
+
+			if (iod->iod_recxs[idx].rx_idx & PARITY_INDICATOR) {
+				skip_list[i][sl_idx++] = -(long)oca->u.ec.e_len;
```

Comment:
I see your point on the sign business.  I wanted to avoid a struct.  I think we need 64 bits, since rx_nr is an unsigned long:

typedef struct {
        /** Indice of the first record in the extent */
        uint64_t        rx_idx;
        /**
         * Number of contiguous records in the extent
         * If \a rx_nr is equal to 1, the extent is composed of a single record
         * at indice \a rx_idx
         */
        uint64_t        rx_nr;
} daos_recx_t;